### DISCUSSION TITLE
[vitest-worker]: Timeout calling "onTaskUpdate"

### DISCUSSION THREAD

**Discussion by vlzuiev** [vitest-dev/vitest#6511] [Q&A]
### Describe the bug

We're facing the following issue when running vitests on CI (Jenkins). 
It's now reproducible locally. All tests are passing. 

Any thoughts on why this issue might appear?

 ⎯⎯⎯⎯ Unhandled Rejection ⎯⎯⎯⎯⎯
 Error: [vitest-worker]: Timeout calling "onTaskUpdate"
  ❯ Object.onTimeoutError ../node_modules/vitest/dist/chunks/rpc.B7Mfb-Yf.js:69:17
  ❯ Timeout._onTimeout ../node_modules/vitest/dist/chunks/index.BpSiYbpB.js:41:41
  ❯ listOnTimeout node:internal/timers:581:17
  ❯ processTimers node:internal/timers:519:7

### Reproduction

The issue is not reproducible locally only on CI.  

### System Info

```shell
AWS ECS agent

vitest configration:

test: {
    environment: 'jsdom',
    globals: true,
    retry: isCi ? 2 : 0,
    setupFiles: './src/setupTests.ts',
    testTimeout: 150_000,
    silent: process.env.CI === 'true'
  }
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by nazarhussain** [comment]
My observation is that RPC timer in `birpc` starts when the test suit is initialized, not when tasks are started executing. I observed that if one suit is queued but rest of the suits take more time than `testTimout` then get following error.

**Comment by jacobshirley** [comment]
Have been experiencing this a lot lately. I use Bazel to execute Vitest. Only happens on CI (Github Actions)

**Comment by greggjs** [comment]
I can reproduce this issue locally when sharding out my tests across multiple terminal sessions and I have found that it occurs more frequently when a shard has a test file that has a large number of tests contained within it (like over 30).

**Comment by maroqijalil** [comment]
I can reproduce the scenario locally under stressed conditions and found that when a test file runs for longer than a certain duration (around 1 minute), the error is triggered.

Splitting the tests in that file into multiple files resolved the issue in my case.

**Comment by kris-ellery** [comment]
I'm seeing it at least 2-3 times per day in GitLab CI. I have 155 test suites with 975 tests. Passes locally w/out any issues on Apple silicon MacBook, but CI is very unreliable. Any ideas on how to fix this?

```
⎯⎯⎯⎯⎯⎯ Unhandled Errors ⎯⎯⎯⎯⎯⎯
Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.
⎯⎯⎯⎯⎯⎯ Unhandled Error ⎯⎯⎯⎯⎯⎯⎯
Error: [vitest-worker]: Timeout calling "onTaskUpdate"
 ❯ Object.onTimeoutError node_modules/.pnpm/vitest@3.1.1_@types+debug@4.1.12_@types+node@22.15.12_@vitest+ui@3.1.1_happy-dom@17.4.6_jiti@_kvzzvg25pb4uej4qh54xdkvdle/node_modules/vitest/dist/chunks/rpc.DGgL5dw7.js:62:10
 ❯ Timeout._onTimeout node_modules/.pnpm/vitest@3.1.1_@types+debug@4.1.12_@types+node@22.15.12_@vitest+ui@3.1.1_happy-dom@17.4.6_jiti@_kvzzvg25pb4uej4qh54xdkvdle/node_modules/vitest/dist/chunks/index.68735LiX.js:55:41
 ❯ listOnTimeout node:internal/timers:[588](RETRACTED/-/jobs/60018627#L588):17
 ❯ processTimers node:internal/timers:523:7
```

**Comment by dubzzz** [comment]
I have a very similar issue on the open-source project fast-check ([here](https://github.com/dubzzz/fast-check/pull/5962)). The issue started to appear when I attempted to bump from 3.0.9 to 3.1.3 and seems to be related to my spec file Docs.spec.ts and its test called file.md. The issue only happens on CI against Ubuntu with Node 22.

**Comment by RichardKaufmannOG** [comment]
> [!CAUTION]
> We also have this issue. Tests seem to work, but afterwards ```Timeout calling "onTaskUpdate"``` appears.
> We use **vitest v3.0.8**.
>
> <img width="1160" alt="image" src="https://github.com/user-attachments/assets/eadd7d47-aed2-49b9-b145-8e28788a0229" />


> [!NOTE]
> Setting higher values for [test-timeout](https://vitest.dev/config/#testtimeout), [hook-timeout](https://vitest.dev/config/#hooktimeout) or [teardown-timeout](https://vitest.dev/config/#teardowntimeout) did not change anything.


> [!TIP]
> **Workaround** that helped us was originally [posted here](https://github.com/vitest-dev/vitest/issues/4497#issuecomment-1887757764). Put this into the test file which was executed right before the error appears:
>
> ```javascript
> afterEach(async () => {
> // already existing afterEach code goes here
> 
>  await new Promise(res => setImmediate(res))
>});
>```

**Comment by kgetz-arista** [comment]
Also running into this after upgrading Vitest from 3.0.9 to 3.1.3. Also seeing failures in CI (Jenkins) only; tests pass just fine on my laptop.

My solution was to break up a large, long-running test file into smaller test files. That seems to have done the trick (no `afterEach` workaround necessary).

**Reply by manuel3108** [reply]
> Also running into this after upgrading Vitest from 3.0.9 to 3.1.3.

Can confirm that downgrading to 3.0.9 fixes it for me as well. All other comments here did not help me resolve this issue

**Comment by velut** [comment]
Setting `pool` to `threads` instead of the default `forks` helps in my case.

```
import { defineConfig } from "vitest/config";

export default defineConfig({
	test: {
		pool: "threads",
	},
});

```

See https://vitest.dev/config/#pool.

**Comment by mikegreiling** [comment]
I am also having this issue. I am using `pool: vmForks`, though switching to `vmThreads` exhibits the same error.

Even if breaking up large tests is a solution, I'd rather not have to limit my test files to arbitrary size limits. My tests should _just work_ even if they are written sub-optimally. Curious what the underlying cause might be.

**Comment by greggjs** [comment]
👋  hey so I thought I'd come back here an update on my situation:

My repository has a ton of emotion based theming, and upon removing the theme provider for emotion, my tests ran 1 second faster **per test**, and I don't have any worker timeout issues anymore. I would like to say the issue was caused by emotion, but it could also be caused by my company's theme provider that uses emotion, so take it with a grain of salt but it worked for us. So, if you can remove unnecessary things for component tests like theme providers from your tests, it will make this issue go away.

**Comment by mikegreiling** [comment]
In my case, I have no indication of which test is the culprit. We are running a suite of 6000+ tests which pass just fine in local development, but in CI we get this very frustrating error message **at the end of the test suite**.
![CleanShot 2025-06-11 at 12 05 12@2x](https://github.com/user-attachments/assets/47eda930-d8cb-47cf-ada6-5a7af587a149)

The stack trace is not useful in narrowing anything down. I do not see any output during the test run. It seems as though something is not cleaned up properly, but I have no idea how to go about diagnosing what's happening.

**Reply by ben-clarke** [reply]
That is possibly because one (or more) of your test files takes longer than 60 seconds to complete. There seems to be a **hardcoded** timeout that throws this error at the end of the whole suite. I am not sure about the reasoning of throwing an error after 60 seconds or if there even is one - it could just be a finger in the air "timeout after 60 seconds" thing.

I would do some analysis on the timings **per file** in the output to see if this the case for you. If it is try breaking it/them down and see if it fixes it.

This is what I have seen in the past - however your case may be different.

**Reply by mikegreiling** [reply]
Thank you, I think that might have done the trick. I took my two longest-running tests and broke them up into smaller files and that got rid of it (at least it seems to, I've only run a few pipelines since)

It would be great if we could get this error message to be a bit more expositive so others don't run into this same headache. A simple `CheckboxTree.test.tsx timed out after 60 seconds` would be very welcome here. If I have time I will try to investigate and throw together a pull request.

**Comment by kinoli** [comment]
I ran into this issue as well. Adding the following to vitest.setup file fixed it.

```
import { flushPromises } from '@vue/test-utils'

afterEach(async () => {
  await flushPromises()
})
```

You may be able to target the actual large tests that are causing it if you want to localize flushPromises more, but I didn't feel I needed to.

**Comment by Tobbe** [comment]
I'm running into this issue as well. Mostly on Windows CI. Never locally. I'm on Vitest 3.2.4

![image](https://github.com/user-attachments/assets/06671edb-bdcf-472c-b6a3-5fb6ecbbaffb)

A lot of the suggestions in this thread is to break larger testfiles up into smaller files. 
My problem is that the test file that fails already only has one test in it, so I can't break it down further.

Running that single test file locally these are the timings: 
![image](https://github.com/user-attachments/assets/d3adff9e-c91f-4617-9a4b-dcc40516db44)

Doesn't seem terribly bad, does it? It is doing some filesystem stuff, so I could look into mocking fs to move more of that to in-memory operations to probably speed it up a little bit more. But I'm not sure that'll actually help fix this issue.

Now trying the `await new Promise((res) => setImmediate(res))` workaround posted by @RichardKaufmannOG. 
https://github.com/cedarjs/cedar/pull/174

**Reply by Tobbe** [reply]
The workaround might have worked for that testcase. But now another one is failing with the same error. 
Would be really nice to understand what the real issue is here

**Reply by Tobbe** [reply]
I was originally putting the workaround in `beforeAll`, but kept seeing this issue in my CI runs. So I switched to `afterEach` as the original workaround did. Still errors out from time to time on CI 🙁  Here's an example https://github.com/cedarjs/cedar/commit/222d16a97c89b650f8731789508e39290cde8e0e

**Comment by AriPerkkio** [comment]
This is fixed in https://github.com/vitest-dev/vitest/releases/tag/v4.0.0-beta.4 by
- https://github.com/vitest-dev/vitest/pull/8297

**Reply by Tobbe** [reply]
Amazing! Thank you so much 🙏

**Reply by tats-u** [reply]
Can it be backported to 3.x? It prevents me from introducing CodSpeed (already adopted by Rolldown etc.).

**Reply by suica** [reply]
Thank you!

**Reply by dubzzz** [reply]
It would be awesome if it could be backported in v3. On [fast-check](https://fast-check.dev/) I struggle into having the CI pass against Windows (most of the runs fail due to "onTaskUpdate").

**Reply by hi-ogawa** [reply]
Any reason you can not switch to v4 beta?

**Reply by dubzzz** [reply]
No real reason except I was waiting for the stable version. But I'll probably switch earlier

**Reply by tats-u** [reply]
This issue is a _regression_ within the same semver range. Is it really acceptable to leave a regression issue unaddressed and force users to use unstable and incomplete versions as guinea pigs?
FYI Java gives higher priorities to regression bugs.

**Reply by tats-u** [reply]
Also I'd like to continue the support for Node 18 for a while. Vitest 4 threw away it. I don't use syntax or features dedicated to Node 20+.

**Reply by jakegregg** [reply]
Node 18 EOL happened earlier this year, so I wouldn't expect to be getting much support for it going forward. 

https://nodejs.org/en/blog/announcements/node-18-eol-support

**Reply by tats-u** [reply]
> We understand that some organizations face constraints that prevent immediate upgrades, such as legacy codebases, compliance requirements, or complex dependency chains. If your company cannot upgrade immediately but needs continued security support for Node.js v18 or earlier versions, [commercial support](https://www.herodevs.com/support/node-nes?utm_source=NodeJS+&utm_medium=Link&utm_campaign=Blog_18_eol_support) is available through HeroDevs.

You can continue to use Node 18 as long as you pay. This is why I do not want to throw away Node 18 immediately in my libraries.
